### ISSUE TITLE
Question: how can use the proof of concept 

### ISSUE THREAD

**Issue by heiko1988** [bertreb/pimatic-home-connect#1]
Hi,

I have installed the plugin in pimatic directly over github. 
I have  type my client ID and secret key. 

No device found from home connect and no debug message in pimatic. 

Can you write a little guide on how to set it up. I know it's a concept for now

**Comment by bertreb** [comment]
Hi, i added a small installation guide and updated the pimatic-home-connect.coffee.

**Comment by heiko1988** [comment]
Thanks.

i have try. pimatic say error error 
`[pimatic]: Cannot find module './lib/utils'`

**Comment by heiko1988** [comment]
ok, i have clone direct in /home/pi/pimatic-app/node_modules/ and then change the main.js, the lib error dosent come again

now say the console log
`[pimatic-home-connect] Home-connect offline`

**Comment by bertreb** [comment]
Ok, i forgot in the instruction that you need to install home-connect-js and lodash first.
In the pimatic-home-connect directory run command -> **npm install**
This will install home-connect-js and lodash
and then replace main.js

**Comment by heiko1988** [comment]
ok, i have new reinstall
now say the console log
`[pimatic-home-connect] Home-connect offline`

no popup come in pimatic


what for date use you in the dev application in home connect?

Access Private
Private access is limited to a single Home Connect user account as defined.
Home Connect user account for testing: here i have my Private acc
OAuth Flow: Device Flow
Client ID: the application id
Client Secret (optional): the applcation secret
Success Redirect ?
One Time Token Mode: Disabled

**Comment by bertreb** [comment]
Its not the device flow but the **Authorisation Code Grant Flow** you need to use for the application.
It will probably change but for the POC its the ACGF

**Comment by bertreb** [comment]
The redirect uri is  http://localhost:3000/o2c

**Comment by bertreb** [comment]
This is all part of the extended instruction 😄

**Comment by heiko1988** [comment]
ok :)

say the same:
`[pimatic-home-connect] Home-connect offline`

is the port 3000? or my pimatic port?
and 
must the port be accessible externally?

**Comment by bertreb** [comment]
its 3000 not your pimatic port

**Comment by heiko1988** [comment]
no i use intern 81 and externally 23003

**Comment by bertreb** [comment]
yes, but you must use 3000. It's not related to the pimatic ports

**Comment by heiko1988** [comment]
ok, then I don't have to change anything? must the port be accessible externally and do I have to change the local host to my ip or my dns?

**Comment by heiko1988** [comment]
ok, i reinstalled everything.
i think:
the error will be that authorization popup does not appear

before the question comes i have no popup blocker or other active things

**Comment by bertreb** [comment]
The plugin runs for authentication a webserver on port 3000 and opens a browser window on it self

**Comment by bertreb** [comment]
A checklist
- cloned the pimatic-home-connect repos
- did a npm install in hte pimatic-home-connect directory
- replaced the main.js
- added and activated the plugin in the config.json
- created the right application in developers Home-Connect
- ClientID and ClientSecret in plugin config
- Restart pimatic -> the authentication browser should showup now

**Comment by bertreb** [comment]
> ok, then I don't have to change anything? must the port be accessible externally and do I have to change the local host to my ip or my dns?

No you don't need to change anything

**Comment by heiko1988** [comment]
Ok, I try again

**Comment by bertreb** [comment]
You need to have the browser on the same computer as pimatic is running on!

**Comment by heiko1988** [comment]
so i have try agian:
this is my way

`root@raspberrypi:~ # cd /home/pi/pimatic-app
root@raspberrypi:/home/pi/pimatic-app # cd ./node_modules
root@raspberrypi:/home/pi/pimatic-app/node_modules # git clone https://github.com/bertreb/pimatic-home-connect
Klone nach 'pimatic-home-connect' ...
remote: Enumerating objects: 119, done.
remote: Counting objects: 100% (119/119), done.
remote: Compressing objects: 100% (83/83), done.
remote: Total 119 (delta 69), reused 81 (delta 33), pack-reused 0
Empfange Objekte: 100% (119/119), 22.52 KiB | 0 bytes/s, Fertig.
Löse Unterschiede auf: 100% (69/69), Fertig.
root@raspberrypi:/home/pi/pimatic-app/node_modules # cd ./pimatic-h
pimatic-harmonyhub/   pimatic-home-connect/
root@raspberrypi:/home/pi/pimatic-app/node_modules # cd ./pimatic-h
pimatic-harmonyhub/   pimatic-home-connect/
root@raspberrypi:/home/pi/pimatic-app/node_modules # cd ./pimatic-home-connect/
root@raspberrypi:/home/pi/pimatic-app/node_modules/pimatic-home-connect # npm install
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated core-js@2.6.11: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.

> core-js@2.6.11 postinstall /home/pi/pimatic-app/node_modules/pimatic-home-connect/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon:
> https://opencollective.com/core-js
> https://www.patreon.com/zloirock

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)

npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN pimatic-home-connect@0.0.2 requires a peer of pimatic@0.9.* but none is installed. You must install peer dependencies yourself.
npm WARN pimatic-home-connect@0.0.2 No license field.

added 153 packages from 137 contributors and audited 280 packages in 42.704s

13 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

root@raspberrypi:/home/pi/pimatic-app/node_modules/pimatic-home-connect #`


then i have the main.js frome here /home/pi/pimatic-app/node_modules/pimatic-home-connect to here /home/pi/pimatic-app/node_modules/pimatic-home-connect/node_modules/home-connect-js

then i type in the config:
```
    {
      "plugin": "home-connect",
    }
```
then start pimatic, and in the plugin i have type the cliendid and secret
   ```
   "plugin": "home-connect",
      "clientId": "xxxxx",
      "clientSecret": "xxxxx",
      "simulation": true,
      "debug": true,
      "active": true
```
and then i have restart, and now come nothin prompt.

what make i wrong ;)

**Comment by heiko1988** [comment]
can make a buttom in plugin conf to start the auth?

**Comment by bertreb** [comment]
In the current setup the browser must run on the same computer Pimatic is running on.
So you can't use a head-less configuration!

**Comment by bertreb** [comment]
For the headless setup (what most a the user use) i need to change the authentication flow to the Device Flow. The home-connect-js  lib doesn't support that and there some work to do.

**Comment by bertreb** [comment]
> can make a buttom in plugin conf to start the auth?

no, thats not a solution (see my 2 posts above)

**Comment by heiko1988** [comment]
ok, that's the problem ;)

I have my pimatic on the pi and try it from my laptop.

then it is clear that it does not work.

is there a way that i can do it via the pc or directly with curl or so on the pi in the console?

**Comment by bertreb** [comment]
No curl will not work. You need to have to login with something like vnc.

**Comment by bertreb** [comment]
But if you have a laptop. Its not very complex to install basic pimatic(+nodejs) and do the install steps for pimatic-home-connect. You start/stop pimatic via commandline.

**Comment by heiko1988** [comment]
Ok, and can this transfer to my pimatic on the pi? After all work?

**Comment by bertreb** [comment]
It could work, i will test that and let you know. This is still a POC way-of-working!

**Comment by bertreb** [comment]
Please remove your ClientId and ClientSecret from the thread :)

**Comment by heiko1988** [comment]
> It could work, i will test that and let you know. This is still a POC way-of-working!

Ok thanks

**Comment by heiko1988** [comment]
> Please remove your ClientId and ClientSecret from the thread :)

I have removed and delete the application on home connect dev page.

**Comment by bertreb** [comment]
Thats not what I ment, but its also a solution 😅 
You need a new application to test

**Comment by heiko1988** [comment]
I am aware of that :)

But safe is safe

**Comment by heiko1988** [comment]
ok, i have check again.

so,

now i have a vnc server on the pi and am connected.

I use the chromium browser and did the whole thing again, including new id and secret.

Unfortunately I can't get it done;)

**Comment by bertreb** [comment]
You are logged in via vnc to the pi? can you open a chrome window in vnc screen?

**Comment by bertreb** [comment]
You need a vnc app on your laptop, you can't use a browser!

**Comment by heiko1988** [comment]
yes i connect per vnc to the pi and in the pi session i open a chromium browser

![Unbenannt](https://user-images.githubusercontent.com/5456599/76170155-91053580-617f-11ea-99f3-80a6582d5d78.PNG)


see is localhost:81 my pimatic

**Comment by bertreb** [comment]
Ok, looks good. What is the log saying?

**Comment by heiko1988** [comment]
the pimatic log say nothing

**Comment by bertreb** [comment]
but there should be information in the deamon.log file on what happened at the startup of Pimatic.

**Comment by heiko1988** [comment]
startup pimatic:

12:26:38.265 [pimatic] info: Loading plugin: "pimatic-home-connect" (0.0.2)

when search device:

12:28:42.556 [pimatic] info: pimatic-home-connect: Searching for new devices
12:28:42.557 [pimatic-home-connect] info: Home-connect offline

tis is the only from home-connect, and no errors from other in the log

**Comment by heiko1988** [comment]
the login promp not comming

**Comment by bertreb** [comment]
can you check in the home-connect-js directory and then the node_modules if fs is installed?

**Comment by bertreb** [comment]
if not then install fs. Go to home-connect-js and execute command npm install fs

**Comment by bertreb** [comment]
These are normally all automatic installs, but in this manual installation and lot of manual patches need to be done. The fs module is probably missing.

**Comment by heiko1988** [comment]
yes that was missing, I installed it with npm i fs and rebooted the vnc session. unfortunately no popup to log in

**Comment by bertreb** [comment]
Please restart pimatic

**Comment by heiko1988** [comment]
unfortunately does not work. I also installed the EventSource.

unfortunately it still doesn't work.

how do you do the restart via console or in plugin and then restart?

**Comment by bertreb** [comment]
You can use ssh in a terminal on your laptop or in the vnc window open a terminal.
The commands are the same sudo service pimatic restart. You can also restart pimatic via the gui (plugin page)

**Comment by heiko1988** [comment]
ok, have already tried both things.

just want to rule out every mistake where I could make

do I have to log out of pimatic?

**Comment by bertreb** [comment]
No you don't have to logout. You are using node v10?

**Comment by heiko1988** [comment]
yes i use : Node.js version 10.16.3

**Comment by bertreb** [comment]
Ok. I'm going to change the authentication to device flow. For that i need to change the home-connect-lib and integrate the functions in the plugin. 
The advantage will be that you get a link/url in the pimatic gui, that is used to authenticate the access. The headless pimatic (like yours) is then supported out-of-the-box. It will take 1-2 days to get a new poc with the device flow.

**Comment by heiko1988** [comment]
That is very nice :)
let the time where you need is in no hurry
I really don't know what I'm doing wrong :)

i'm looking forward to something new to test

**Comment by bertreb** [comment]
I've changed the api library and published on npm version 0.0.4.
This should make installation easier 😄 
See the readme for instructions

**Comment by heiko1988** [comment]
Hi,

Thanks :) 

The plugin work fine with the simulation. When I start a sim on the home connect page, I see all in pimatic very fine. 

1 problem I have. When I have all finished and work. I make a restart then come this 

`14:20:17.064 [pimatic] New device "CoffeeMaker Simulator"...
14:20:25.553 [pimatic-deebot] Cleanings 1
14:20:27.887 [pimatic-echo] started hue emulator on port 80
14:20:27.923 [pimatic-sysinfo] Mounted File Systems: /, /boot
14:20:27.932 [pimatic] An uncaught exception occurred: TypeError: Cannot read property 'getAppliance' of null
14:20:27.932 [pimatic]>    at Timeout._onTimeout (/home/pi/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:244:29)
14:20:27.932 [pimatic]>    at ontimeout (timers.js:436:11)
14:20:27.932 [pimatic]>    at tryOnTimeout (timers.js:300:5)
14:20:27.932 [pimatic]>    at listOnTimeout (timers.js:263:5)
14:20:27.932 [pimatic]>    at Timer.processTimers (timers.js:223:10)
14:20:27.932 [pimatic]> This is most probably a bug in pimatic or in a module, please report it!
14:20:27.934 [pimatic] exiting...`


Pic
![Screenshot_20200310-140057_JuiceSSH](https://user-images.githubusercontent.com/5456599/76316707-50baca00-62db-11ea-9852-c82c3e1c5a31.jpg)

**Comment by bertreb** [comment]
Hi, i did a small adjustment in version 0.0.5 for this error.
But the plugin was working and when came this error?

**Comment by heiko1988** [comment]
Exactly, the plugin itself works. 

The data from the simulation are displayed in pimatic.

 Only a restart is not possible.

 Ps: I'll update right away and give you feedback

**Comment by heiko1988** [comment]
Hi,

Saim fail with same error description. 

When I delete the simulation device in the config.json then pimatic starting normally with no errors.

**Comment by bertreb** [comment]
you created the HomeconnectManager device and used the Grant Flow credentials in that device?

**Comment by bertreb** [comment]
If you have done that without any homeconnect devices, what happens if you restart?

**Comment by bertreb** [comment]
I added an extra check on restart. The function that automatically starts a device after discovery is probably causing the error at restart. -> version 0.0.6

**Comment by heiko1988** [comment]
> you created the HomeconnectManager device and used the Grant Flow credentials in that device?

Yes is it, and the home connect manager device is created wit id and secret




Ok I update and check again

**Comment by heiko1988** [comment]
Thank you, 
Now the simulation work fine 
And restart work :+1:

Can I use only for simulation? Or physically coffeemaker 

On the basis I also have a small project. 
So that I get my coffee fully automatically in the morning 

A small motor and a treadmill. When the coffee machine starts, the rinsing water goes into one cup and the coffee into the other

**Comment by bertreb** [comment]
Thats nice. I need to add something, because the logic is that you are connected to the simulator or to the real world. When you have already simulator devices and you are changing to real world and back, the sim or the real devices need to be disabled.

**Comment by bertreb** [comment]
I'm moving the selection of - simulation vs live - back to the plugin level.
Thats much clearer and the only disadvantage is the restart thats needed when you switch between live and simulator

**Comment by heiko1988** [comment]
Yes, that's completely ok. You don't usually switch between simulation and real so often

**Comment by bertreb** [comment]
I published version 0.0.8 with the credentials and simulator choice  in the plugin config.
When you enable Live (= disable simulator) in the HomeconnectManager the authorisation url will be under the **Link** (popup of the label)
I can't test the live mode. So let me know 😅 and its just reading status, no actions yet!

**Comment by heiko1988** [comment]
Year the physical coffe machine work very fine

![Screenshot_20200310-171605_Chrome Beta](https://user-images.githubusercontent.com/5456599/76334216-034a5700-62f3-11ea-864e-c7bf91393d2e.jpg)


Is there a status for the rinse program? 
![Screenshot_20200310-171529_Chrome Beta](https://user-images.githubusercontent.com/5456599/76335891-7ead0800-62f5-11ea-978f-ea9e9ae2271b.jpg)

That would be important for my project, the progress in percent can only be seen in the program.

Ps: the buttons who see of the pic is over ifft

**Comment by bertreb** [comment]
The CoffeeMaker, Washer, Oven and Dishwasher are supported. I can easily add the rest.
But the device you having is also determining what the options and statuses are.
The progress for the Washer is available (if that is what you mean by rinse)

**Comment by bertreb** [comment]
what would be logical actions to have? Buttons can be easily created from that

**Comment by heiko1988** [comment]
The Google translated, it does not think so in a nutshell;) So when I start the coffee machine, it does a cleaning with normal water. Then I can start a coffee The start process is called normal in German spülen or reinigen

**Comment by heiko1988** [comment]
Button Kaffee an Start the coffee machine
Button kaffe aus turn off the coffee machine 
Kaffe schwarz start a coffe black, With the settings where and the machine are already set
Button latte start a coffe latte, With the settings where and the machine are already set

Button Telegramm is a test by me to notify in telegram ;)

**Comment by bertreb** [comment]
Ok i understand, the rinse is not something i found in the api, but i'll check
I'll test the commands

**Comment by heiko1988** [comment]
I have a error ;) 
Come every second

18:52:48error [pimatic-home-connect]: Terminated events stream for all appliances: Home Connect API error: Gateway Timeout

Then I have restarted pimatic no errors come more , now status of home connect manager is initialisiert. And become no info to the coffee maker. See pic

![Screenshot_20200310-190718_Chrome Beta](https://user-images.githubusercontent.com/5456599/76344752-6394c500-6302-11ea-918a-c0d972a7d716.jpg)


Ok
I have add a new device home connect Manger, and restart and all work :)

**Comment by heiko1988** [comment]
Ok. After I add a new device home connect Manger and make a restart. Work all normally.

**Comment by heiko1988** [comment]
And give the stat out of beans and water is empty?

**Comment by bertreb** [comment]
In the CoffeeMaker Device gui the beans is called Strength and water is called FillQuantity

**Comment by heiko1988** [comment]
ok,

i have check on the coffeemaschine the FillQuantity say 
how many ml should be in the cup

Say nothing water empty 
![20200310_200457](https://user-images.githubusercontent.com/5456599/76349843-0cdfb900-630b-11ea-88cc-14c878e18156.jpg)

And 

![Screenshot_20200310-200445_Chrome Beta](https://user-images.githubusercontent.com/5456599/76349879-1b2dd500-630b-11ea-9963-25384cc5d95c.jpg)

**Comment by bertreb** [comment]
Thats correct i think. In the simulator you can set it between 100 and 300

**Comment by bertreb** [comment]
besides starting the program you to set the Strength and the fillQuantity?

**Comment by heiko1988** [comment]
I don't think we're talking about the same :)

i mean message water tank empty in pimatic nothin see

**Comment by heiko1988** [comment]
> besides starting the program you to set the Strength and the fillQuantity?

i make this direct on the coffe maschine, and only start the programm over pimatic in this time.

if you integrate it then I can also do it via pimatic;)

but that is not so important because I mostly use the same coffee settings anyway

**Comment by bertreb** [comment]
You are talking about the status of the water reservoir and the beans?

**Comment by heiko1988** [comment]
Exactly, that's what I mean

**Comment by bertreb** [comment]
I didn't found any status info on beans and water. Is the app providing this info ?
In version 0.0.10 you can do the basic commands start and stop. Pause and resume is not supported on any of the simulator devices.

**Comment by heiko1988** [comment]
> I didn't found any status info on beans and water. Is the app providing this info ?
> In version 0.0.10 you can do the basic commands start and stop. Pause and resume is not supported on any of the simulator devices.

In the home connect app and In ifft becomes this info for beans and water.

The iobroker home connect plugin have this. https://github.com/dna909/ioBroker.homeconnect

Maybe you can find something here in the source code


Thanks for updating, I test it.

**Comment by bertreb** [comment]
I added the supported for events status of the supported devices. So BeanContainerEmpty and WaterTankEmpty could work and even DripTrayFull
Can't test it because because the simulator doesn't support this, but i will hear if this is working.
Version 0.0.14

**Comment by heiko1988** [comment]
Ok. Thanks i try this later. And give here a feedback.

Update:
Unfortunately I didn't make it today. To get the container full ;) I will contact you as soon as I have made it. And work or not. 

Thanks again for your efforts and that you are always so fast

**Comment by heiko1988** [comment]
so i have update:

WaterTankEmpty  and  DripTrayFull.

become this error:

`
error [pimatic-home-connect]: Terminated events stream for all appliances: Parse Error
15:35:14error [pimatic]: An uncaught exception occurred: TypeError: Cannot set property 'value' of undefined
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:400:11)
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:334:18)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeConnectAPI.<anonymous> (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:318:14)
    at HomeConnectAPI.emit (events.js:198:13)
    at dispatch (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:710:51)
    at requestStream.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:744:56)
    at chunk.split.forEach.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:58)
    at Array.forEach (<anonymous>)
    at IncomingMessage.response.on.chunk (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:42)
    at IncomingMessage.emit (events.js:203:15)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:265:13)
    at IncomingMessage.Readable.push (_stream_readable.js:224:10)
    at HTTPParser.parserOnBody (_http_common.js:124:22)
    at TLSSocket.socketOnData (_http_client.js:451:20)
    at TLSSocket.emit (events.js:198:13)
    at TLSSocket.emit (/root/pimatic-app/node_modules/pimatic/startup.coffee:64:24)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at TLSSocket.Readable.push (_stream_readable.js:224:10)
    at TLSWrap.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
 This is most probably a bug in pimatic or in a module, please report it!
15:35:13error [pimatic-home-connect]: Terminated events stream for all appliances: Parse Error
15:35:13error [pimatic]: An uncaught exception occurred: TypeError: Cannot set property 'value' of undefined
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:400:11)
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:334:18)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeConnectAPI.<anonymous> (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:318:14)
    at HomeConnectAPI.emit (events.js:198:13)
    at dispatch (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:710:51)
    at requestStream.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:744:56)
    at chunk.split.forEach.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:58)
    at Array.forEach (<anonymous>)
    at IncomingMessage.response.on.chunk (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:42)
    at IncomingMessage.emit (events.js:203:15)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:265:13)
    at IncomingMessage.Readable.push (_stream_readable.js:224:10)
    at HTTPParser.parserOnBody (_http_common.js:124:22)
    at TLSSocket.socketOnData (_http_client.js:451:20)
    at TLSSocket.emit (events.js:198:13)
    at TLSSocket.emit (/root/pimatic-app/node_modules/pimatic/startup.coffee:64:24)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at TLSSocket.Readable.push (_stream_readable.js:224:10)
    at TLSWrap.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
 This is most probably a bug in pimatic or in a module, please report it!
15:35:12error [pimatic-home-connect]: Terminated events stream for all appliances: Parse Error
15:35:08error [pimatic]: An uncaught exception occurred: TypeError: Cannot set property 'value' of undefined
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:400:11)
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:334:18)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeConnectAPI.<anonymous> (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:318:14)
    at HomeConnectAPI.emit (events.js:198:13)
    at dispatch (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:710:51)
    at requestStream.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:744:56)
    at chunk.split.forEach.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:58)
    at Array.forEach (<anonymous>)
    at IncomingMessage.response.on.chunk (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:42)
    at IncomingMessage.emit (events.js:203:15)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:265:13)
    at IncomingMessage.Readable.push (_stream_readable.js:224:10)
    at HTTPParser.parserOnBody (_http_common.js:124:22)
    at TLSSocket.socketOnData (_http_client.js:451:20)
    at TLSSocket.emit (events.js:198:13)
    at TLSSocket.emit (/root/pimatic-app/node_modules/pimatic/startup.coffee:64:24)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at TLSSocket.Readable.push (_stream_readable.js:224:10)
    at TLSWrap.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
 This is most probably a bug in pimatic or in a module, please report it!
15:35:07error [pimatic-home-connect]: Terminated events stream for all appliances: Parse Error
15:35:03error [pimatic]: An uncaught exception occurred: TypeError: Cannot set property 'value' of undefined
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:400:11)
    at HomeconnectDevice.getProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:334:18)
    at HomeconnectDevice.setProgramOrOption (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:1:1)
    at HomeConnectAPI.<anonymous> (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:318:14)
    at HomeConnectAPI.emit (events.js:198:13)
    at dispatch (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:710:51)
    at requestStream.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:744:56)
    at chunk.split.forEach.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:58)
    at Array.forEach (<anonymous>)
    at IncomingMessage.response.on.chunk (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:42)
    at IncomingMessage.emit (events.js:203:15)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:265:13)
    at IncomingMessage.Readable.push (_stream_readable.js:224:10)
    at HTTPParser.parserOnBody (_http_common.js:124:22)
    at TLSSocket.socketOnData (_http_client.js:451:20)
    at TLSSocket.emit (events.js:198:13)
    at TLSSocket.emit (/root/pimatic-app/node_modules/pimatic/startup.coffee:64:24)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at TLSSocket.Readable.push (_stream_readable.js:224:10)
    at TLSWrap.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
 This is most probably a bug in pimatic or in a module, please report it!
`

**Comment by bertreb** [comment]
I saw the BIG error message and released an update to catch that error. Can you post the debug message?

**Comment by heiko1988** [comment]
there is nothing in the debug log, only got that as an error

**Comment by bertreb** [comment]
did you install version 0.0.15? On the line the error occurred, i placed a try/catch to get the info in a debug message.

**Comment by heiko1988** [comment]
yes i have updatet to 0.0.15
no in the debug log

only the error come on startup the coffe maschine.

`error [pimatic-home-connect]: Terminated events stream for all appliances: Parse Error
20:22:11error [pimatic]: An uncaught exception occurred: TypeError: Cannot read property 'length' of undefined
    at HomeConnectAPI.<anonymous> (/root/pimatic-app/node_modules/pimatic-home-connect/pimatic-home-connect.coffee:325:11)
    at HomeConnectAPI.emit (events.js:198:13)
    at dispatch (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:710:51)
    at requestStream.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:744:56)
    at chunk.split.forEach.line (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:58)
    at Array.forEach (<anonymous>)
    at IncomingMessage.response.on.chunk (/root/pimatic-app/node_modules/pimatic-home-connect/homeconnect_api.js:834:42)
    at IncomingMessage.emit (events.js:203:15)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:265:13)
    at IncomingMessage.Readable.push (_stream_readable.js:224:10)
    at HTTPParser.parserOnBody (_http_common.js:124:22)
    at TLSSocket.socketOnData (_http_client.js:451:20)
    at TLSSocket.emit (events.js:198:13)
    at TLSSocket.emit (/root/pimatic-app/node_modules/pimatic/startup.coffee:64:24)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at TLSSocket.Readable.push (_stream_readable.js:224:10)
    at TLSWrap.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
 This is most probably a bug in pimatic or in a module, please report it!`

**Comment by bertreb** [comment]
I added more error handling in version 0.0.17
The problem is i can't test the BeanEmpty, etc events, so you are the tester
I need the eventdata, that is emitted when the Water, Bean and Tripping events occur

**Comment by heiko1988** [comment]
Ok. I have updated to 0.0.17

On startup the Maschine no errors and nothing in debug log. 


I can give you tomorrow my login details for My home connect or we can work over teamviewer. Then can you check all you need. And I can physically empty water, beans etc.

**Comment by bertreb** [comment]
yes that could be options to speed up the debugging.
I changed also a bit the handling of the events. thats could be a solution.
Let me know when you have time tomorrow to choose the approach for debugging.
And lets move to the chat.

**Comment by heiko1988** [comment]
Ok. Thanks.

I wirte you tomorrow
